### ISSUE TITLE
Invalid memory access on JSON deserialization

### ISSUE THREAD

**Issue by j8r** [crystal-lang/crystal#8613]
The following code produces an `Invalid memory access`:
```cr
require "json"

JSON_DATA = <<-E
{
  "class": "task",
  "metadata": {
    "container_progress": ""
  }
}
E

struct Obj
  include JSON::Serializable

  getter class : String
  @metadata : Nil
end

Obj.from_json JSON_DATA
```
I have found 2 ways (quick fixes) to not have `Invalid memory access`:
- don't use `getter class` macro, use `@class` instead, but an exception will be raised.
- fix the `@metadata` type, which is wrong.

To sum-up, having `getter class` along with a wrong type annotation make the bug happen.
Redefining the `#class` method with `def class; @class; end` produces the bug too.

Forbidding to overwrite the `#class` method can be a solution.

**Comment by straight-shoota** [comment]
It should probably be forbidden to override `#class` and `#crystal_type_id`. I don't think there could be a valid use case for overriding these methods.

**Comment by j8r** [comment]
Here the use case was that the REST API returns a `"class"` field.
We will have to use `@[JSON::Field(key: "class_field")]`, which can be fine.

**Comment by straight-shoota** [comment]
Yes, But naming that property `class` interferes with Crystal internals where `#class` method is supposed to return the type of an object.
